website/docs/tstruct.md

```diff
@@ -124,11 +191,286 @@ class ChildTwo < T::Struct
+
+  - There are no static type checks.
+  - Required properties missing in the `Hash` **do** cause raise exceptions at
+    runtime.
+  - The types provided via the `Hash` are **not** checked at runtime.
+
+- Because [union-typed](union-types.md) properties containing `T::Struct`
+  instances are not serialized, they must also not be still serialized when
+  given to `from_hash`:
+
+  ```ruby
+  class Foo < T::Struct
+    prop :foo, Integer
+  end
+  class Bar < T::Struct
+    prop :bar, String
+  end
+
+  class Top < T::Struct
+    prop :foo_or_bar, T.any(Foo, Bar)
+  end
+
+  foo = Foo.new(foo: 12)
+  p(Top.from_hash({"foo_or_bar" => foo}))
+  # => <Top foo_or_bar=<Foo foo=12>>
+  p(Top.from_hash({"foo_or_bar" => foo.serialize}))
+  # => <Top foo_or_bar={"foo"=>12}>
+  ```
+
+  And since there are no runtime type checks, the serialized hash value is
+  directly set to the `foo_or_bar` field.
+
+## Structural vs reference equality
+
+By default, `T::Struct` values compare using reference equality ("Are these two
+instances literally the same object in memory?""), while classes created with
+Ruby's `Struct` class compare using structural equality ("Are these two
+possibly-different objects both instances of the same class, containing
+pairwise-equal fields?").
+
+While it would be nice if `T::Struct` had been built from the beginning with
+structural equality, it wasn't, and now quite a lot of code in the wild depends
+on this.
+
+For those cases where structural equality is preferred, we recommend defining a
+custom module that can be included into a `T::Struct` to override the equality
+methods, providing structural equality.
+
+## Immutable property updates using `with`
+
+Properties defined with `const` do not have setter methods, making it impossible
+to update these properties after construction. A common pattern when working
+with such classes is to "immutably update" the instance by creating a copy of an
+object with identical fields except with a different value for the one `const`
+property.
+
+This is built into `T::Struct`, but has some limitations:
+
+```ruby
+class A < T::Struct
+  const :foo, Integer
+  const :another_required, Integer
+end
+
+a1 = A.new(foo: 1, another_required: 42)
+p(a1) # => <A foo=1 another_required=42>
+
+# The `with` method
+a2 = a1.with(foo: 2)
+p(a2) # => <A foo=2 another_required=42>
+```
+
+Added in haste, the implementation of `with` uses `from_hash` to merge the new
+and old properties and create the new instance. This means it suffers from
+exactly the same gotchas mentioned in the
+[`from_hash` gotchas](#from_hash-gotchas) section above.
+
+## Legacy and Stripe-specific options
+
+There are a number of other legacy or Stripe-internal options in the `prop` DSL.
+Those include `enum`, `foreign`, `ifunset`, `immutable`, `raise_on_nil_write`,
+`redaction`, and `sensitivity`. Stripe employees can reference
+[these docs](http://go/chalk-odm-docs) to learn more.
```

Comment:
Does `dont_store` also fall under here?

Reply:
Yes, thanks, I will add it to this list.